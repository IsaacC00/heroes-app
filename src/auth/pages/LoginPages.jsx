import { useContext } from "react";
import { useNavigate } from "react-router-dom"
import { AuthContext } from "../context/AuthContext";

export const LoginPages = () => {

  //?lo que basicamente hacemos aqui es redireccionar al usuario cuenado se logee
  const { login } = useContext(AuthContext);
  const navigate = useNavigate();

  const onLogin = () => {

    //?regresar al ultimo path en el cual cerre sesion
    //? si es null entonces vamos al inicio de heroes
    const lastPath = localStorage.getItem('lastPath') || '/' //?seteamos el path en PrivbateRoute


    login('Isaac Calderon');
    navigate(lastPath,{
      replace:true
    })
  }

  return (
    <>
      <div className='container mt-5'>

        <h1>Login Page</h1>
        <hr />
        <button
          className='btn btn-primary'
          onClick={onLogin}
        >
          Login
        </button>

      </div>
    </>
  )
}
