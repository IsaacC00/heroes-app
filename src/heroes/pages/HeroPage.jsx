import { Navigate, useNavigate, useParams } from "react-router-dom"
import { getHeroByid } from "../helpers";
import { useMemo } from "react";

export const HeroPage = () => {

  const { id } = useParams();

  //?memorizar el componenete para no ocupar 
  //? bastante recursos en memoria 
  //? cada que el id cambie para que no lo redibuje otra vez 
  //? ojo recordamos el callback
  const hero = useMemo(() => getHeroByid(id), [id]);
  
  //? en caso de no encontrar el heroe 
  //? debemos de redireccionar al usuario 
  //? a otrra pantalla 
  if (!hero) {
    return <Navigate to='/search' />;
  }
  
  //? regresar hacia atras
  //?regresar para aptra utilizando el navigate
  const navigate = useNavigate()
  const onNavigateBack = () => {
    //? regresar un paso anterior
    navigate(-1)
  }

  return (
    <div className="row mt-5 animate__animated animate__fadeInLeft">
      
      <div className=" col-4">
        <img src={`/assets/${id}.jpg`} alt={hero.superhero} className="img-thumbnail" />
      </div>

      <div className="col-8">
        <h3>{hero.superhero}</h3>
        <ul className="list-group list-group-flush">
          <li className="list-group-item"><b>Alter Ego</b> {hero.alter_ego}</li>
          <li className="list-group-item"><b>Publisher</b> {hero.publisher}</li>
          <li className="list-group-item"><b>First Appearance</b> {hero.first_appearance}</li>
        </ul>
        <h5 className="mt-3">Characters</h5>
        <p className="">{hero.characters}</p>
        <button 
          onClick={onNavigateBack}
        className="btn btn-outline-primary"
        >
          Regresar
        </button>
      </div>

    </div>

  )
}
