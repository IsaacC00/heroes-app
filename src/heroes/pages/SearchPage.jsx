import { useLocation, useNavigate } from 'react-router-dom';
import queryString  from 'query-string';
import { HeroCard } from '../components';
import { useForm } from '../../hooks';
import { getHeroesByName } from '../helpers';


export const SearchPage = () => {
 
  
  const navigate = useNavigate();
  //? utilizamos locacion para
  //? recuperar el query de nuestra url
  const location = useLocation();
  
  //? pero debemos instalar query-string  
  //? para poder obtener el valor exacto
  const {q = ''} = queryString.parse(location.search);
  
  
  //?buscamos el heroe
  const heroes = getHeroesByName(q);
  
  const { searchText, onInputChange, } = useForm({
    searchText:q,
  });

  //?condiciones para mostrar compoenentes
  const showSearch = (q.length === 0);
  const showError = (q.length > 0 )&& heroes.length === 0;
  
  //? utilizamos nuestro cuustom hook de formulario
  const onSearchSubmit = (event) => {
    event.preventDefault();
    
    // if (searchText.trim().length <= 1 ) return;
    //? pasamos como argumento el texto que estamos buscando
    //? del formiulario para encontrar el heroe
    navigate( `?q=${searchText}` );
    
  }

  return (
    <>
      <div className="row">

        <h1>Busqueda</h1>
        <hr />

        <div className="col-5">
          <h4>Searching...</h4>
          <hr />
          <form aria-label='formHero' onSubmit={ onSearchSubmit } action="">
            <input
              type="text"
              placeholder="Search Hero"
              className="form-control"
              name="searchText"
              value={searchText}
              onChange={onInputChange}
              autoComplete="off"
            />
            <button className="btn btn-outline-primary mt-2">
              Search
            </button>
          </form>
        </div>
        <div className="col-7">
            <h4>Results</h4>
            <hr />
            
            <div className='alert alert-primary animate__animated animate__fadeIn'
            style={ {display : ` ${showSearch? '' : 'none'} ` } }
            >
              Search a Hero
            </div>
            
            <div  aria-label='divSearch'className='alert alert-danger animate__animated animate__fadeIn'
            style={ {display : `${showError? '' : 'none'}` } }
            >
              No hero with <b>{q}</b>
            </div>
              {/*//? utilizamos un .map por que devolvemos un arreglo de nuestra funcion getHeroesByName */}
              { heroes.map( hero => (
              
              <HeroCard key={hero.id} {...hero} />

                ))
              } 
        </div>
      </div>
    </>
  )
}
