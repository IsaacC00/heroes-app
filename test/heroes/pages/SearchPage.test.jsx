import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { SearchPage } from "../../../src/heroes/pages/SearchPage";

const mocked = jest.fn();

jest.mock('react-router-dom',() => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mocked
})) ;

describe('Pruebas en SearchPage,jsx', () => {

    beforeEach( ()=>jest.clearAllMocks() );

    test('debe de mostar los valores por defecto', () => {

        const { container } = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        );

        expect(container).toMatchSnapshot();

    });

    test('debe de mostar el valor de batman junto con el valor de queryString', () => {

        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        expect(input.value).toBe('batman');

        const img = screen.getByRole('img');
        expect(img.src).toContain('/assets/dc-batman.jpg');

        const alert = screen.getByLabelText('divSearch');
        expect(alert.style.display).toBe('none')

    });

    test('debe mostrar un error si no se encuentra el heroe', () => {

        render(
            <MemoryRouter initialEntries={['/search?q=asdff']}>
                <SearchPage />
            </MemoryRouter>
        );

        const alert = screen.getByLabelText('divSearch');
        expect(alert.style.display).toBe('')

    });

    test('debe llamar el navigate a la pantalla nueva ', () => { 
        
        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        fireEvent.change(input,{target:{name:'searchText',value:'superman'}})
        
        const form = screen.getByLabelText('formHero');
        fireEvent.submit(form);

        expect(mocked).toHaveBeenCalledWith('?q=superman');
        


     });

});