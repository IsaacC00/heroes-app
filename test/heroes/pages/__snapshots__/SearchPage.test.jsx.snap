// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en SearchPage,jsx debe de mostar los valores por defecto 1`] = `
<div>
  <div
    class="row"
  >
    <h1>
      Busqueda
    </h1>
    <hr />
    <div
      class="col-5"
    >
      <h4>
        Searching...
      </h4>
      <hr />
      <form
        action=""
        aria-label="formHero"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search Hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-2"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Search a Hero
      </div>
      <div
        aria-label="divSearch"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        No hero with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
